evm/src/arithmetic/addcc.rs

```diff
@@ -0,0 +1,384 @@
+    }
+    cy
+}
+
+fn eval_ext_circuit_pol_eval_equal<F, const D: usize, I, J>(
+    builder: &mut plonky2::plonk::circuit_builder::CircuitBuilder<F, D>,
+    yield_constr: &mut RecursiveConstraintConsumer<F, D>,
+    filter: ExtensionTarget<D>,
+    pol_coeffs: I,
+    digits: J,
+    is_two_row_op: bool,
+) -> ExtensionTarget<D>
+where
+    F: RichField + Extendable<D>,
+    I: Iterator<Item = ExtensionTarget<D>>,
+    J: Iterator<Item = ExtensionTarget<D>>,
+{
+    // 2^LIMB_BITS in the base field
+    let overflow_base = F::from_canonical_u64(1 << LIMB_BITS);
+    // 2^LIMB_BITS in the extension field as an ExtensionTarget
+    let overflow = builder.constant_extension(F::Extension::from(overflow_base));
+    // 2^-LIMB_BITS in the base field.
+    let overflow_inv = F::inverse_2exp(LIMB_BITS);
+
+    let mut cy = builder.zero_extension();
+    for (a, b) in pol_coeffs.zip_eq(digits) {
+        // t0 = cy + a
+        let t0 = builder.add_extension(cy, a);
+        // t  = t0 - b
+        let t = builder.sub_extension(t0, b);
+        // t1 = overflow - t
+        let t1 = builder.sub_extension(overflow, t);
+        // t2 = t * t1
+        let t2 = builder.mul_extension(t, t1);
+
+        let filtered_limb_constraint = builder.mul_extension(filter, t2);
+        if is_two_row_op {
+            yield_constr.constraint_transition(builder, filtered_limb_constraint);
+        } else {
+            yield_constr.constraint(builder, filtered_limb_constraint);
+        }
+
+        cy = builder.mul_const_extension(overflow_inv, t);
+    }
+    cy
+}
+
+fn eval_packed_generic_check_is_one_bit<P: PackedField>(
+    yield_constr: &mut ConstraintConsumer<P>,
+    filter: P,
+    x: P,
+) {
+    yield_constr.constraint(filter * x * (x - P::ONES));
+}
+
+fn eval_ext_circuit_check_is_one_bit<F: RichField + Extendable<D>, const D: usize>(
+    builder: &mut plonky2::plonk::circuit_builder::CircuitBuilder<F, D>,
+    yield_constr: &mut RecursiveConstraintConsumer<F, D>,
+    filter: ExtensionTarget<D>,
+    x: ExtensionTarget<D>,
+) {
+    let constr = builder.mul_sub_extension(x, x, x);
+    let filtered_constr = builder.mul_extension(filter, constr);
+    yield_constr.constraint(builder, filtered_constr);
+}
+
+/// Constrains x + y == z + cy*2^256 assuming filter != 0.
+pub(crate) fn eval_packed_generic_add_cc<P: PackedField>(
+    yield_constr: &mut ConstraintConsumer<P>,
+    filter: P,
+    x: &[P],
+    y: &[P],
+    z: &[P],
+    cy: P,
+    is_two_row_op: bool,
+) {
+    debug_assert!(x.len() == N_LIMBS && y.len() == N_LIMBS && z.len() == N_LIMBS);
+
+    let pol_sum = x.iter().zip(y).map(|(&xi, &yi)| xi + yi);
+    let expected_cy = eval_packed_generic_pol_eval_equal(
+        yield_constr,
+        filter,
+        pol_sum,
+        z.iter().copied(),
+        is_two_row_op,
+    );
+    // We don't need to check that expected_cy is 0 or 1, since cy has
+    // already been checked to be 0 or 1.
+    if is_two_row_op {
```

Comment:
In which case does `is_two_row_op==true` when this is called? Isn't addition/subtraction/lt/gt a one row op?

Reply:
It is [used in `modular`](https://github.com/mir-protocol/plonky2/blob/hamish/unify-add-sub-cmp-operations/evm/src/arithmetic/modular.rs#L352) to verify that the result is reduced wrt the modulus. Modular operations are two rows and the inputs for this verification are spread across both rows, hence the need for `.constraint_transition(...)` rather than plain `.constraint(...)`. It's a somewhat ugly solution; happy to consider other options!

Reply:
Ah my bad, ignore my comment then.